Blog: add Jesse FROST post (#37)
bitcoinbrink/website: _posts/2021-04-15-jesse-frost.md

---

> title: "Jesse Posner on Flexible Round-Optimized Schnorr Threshold Signatures (FROST)"

Reviewer: Maybe switch the title and description:

title: "Jesse Posner on FROST"
description: "Jesse Posner explains the Flexible Round-Optimized Schnorr Threshold Signatures (FROST) signature scheme."

So the title is pithy and the description adds more detail.

permalink: /blog/2021/04/15/frost/
layout: post
author: jesse
category: "technical"
description: "Jesse Posner explains the FROST signature scheme."
---

The promise of Bitcoin is only as secure as the cryptographic keys that
authenticate its operations: they are the keys to freedom in the digital age.
However, the requirements for Bitcoin key management are becoming more demanding
with innovative off-chain protocols such as the [Lightning Network][ln] and
[Discreet Log Contracts][dlc] that introduce new roles and responsibilities for
keys. In addition, Bitcoin's relentless rise in value creates a need for

> sophisticated systems that can help make rubber-hose cryptanalysis uneconomical.

Reviewer: I'm not sure how widely understood "rubber-hose cryptanalysis" will be to a general audience. Perhaps use more explicit language, or link to https://en.wikipedia.org/wiki/Rubber-hose_cryptanalysis?

Off-chain protocols facilitate smart contracts that are highly scalable and
private, which enable a vast array of new markets and financial instruments.
However, many of these protocols present challenges from a key management
perspective because they rely on “hot” keys that must be readily accessible and
frequently reused, unlike safer "cold" keys that are rarely accessed and never
reused. With the introduction of [Taproot][bip341], we can begin to bridge this
divide between a "hot" and "cold" key with distributed key generation and
threshold signing.

> Distributed key generation uses secure multi-party computation ([MPC][mpc wiki])
> to generate a pair of public and private keys with multiple parties such that no
> party learns the private key. Instead, each party learns a "share" of the

Reviewer: This sentence could be slightly reformulated to make it clearer. "a pair of public and private keys" is slightly ambigious since it could mean "a pair of key pairs", and "with multiple parties" is separated from the thing it's refering to ("secure multi-party computation") and might be parsed as refering to "private keys".

Perhaps something like:

```suggestion
In distributed key generation, multiple parties use secure multi-party computation ([MPC][mpc wiki])
to generate a single public/private key pair, such that no
party learns the private key. Instead, each party learns a "share" of the
```

private key, and when a threshold of these parties cooperate they can use their
shares to generate signatures for the key. The threshold is configured by the
parties when the key is generated. For example, if 5 parties generated a key
with a threshold of 3, then at least 3 of the parties would need to cooperate to
produce a signature (i.e. 3-of-5).

This means that for the entire lifecycle of the private key–from generation to
expiration–it is never constructed or represented on any device in any form

> (assuming a threshold of shares has not been compromised). With Taproot, we can
> leverage the linearity of Schnorr signatures to perform signing with 3
> communication rounds (or 2 rounds with limited concurrency), unlike the ECDSA
> variants that are complex and require many more rounds, and, with the recent
> introduction of Flexible Round-Optimized Schnorr Threshold Signatures
> ([FROST][frost whitepaper]), signing can now be done in a single round with
> unlimited concurrency.

Reviewer: "With Taproot, ... with unlimited concurrency." is a very long sentence. Break it up:

```suggestion
(assuming a threshold of shares has not been compromised). With Taproot, we can
leverage the linearity of Schnorr signatures to perform signing with 3
communication rounds (or 2 rounds with limited concurrency). This is in contrast
to ECDSA multisignature schemes, which are complex and require many more rounds. With the recent
introduction of Flexible Round-Optimized Schnorr Threshold Signatures
([FROST][frost whitepaper]), signing can even be done in a single round with
unlimited concurrency.
```

FROST is a relatively new scheme yet some notable open-source implementations

> have already been developed, such as [isislovecruft/frost-dalek][frost dalek github],
> a Rust implementation using the Ristretto group,
> [taurusgroup/frost-ed25519][frost ed25519 github], a Go
> implementation using Ed25519, and [ZcashFoundation/redjubjub][redjubjub github], a Rust

Reviewer: perhaps remove the github usernames/orgs so these are just refering to the names of the projects:

```suggestion
have already been developed, such as [frost-dalek][frost dalek github],
a Rust implementation using the Ristretto group,
[frost-ed25519][frost ed25519 github], a Go
implementation using Ed25519, and [redjubjub][redjubjub github], a Rust
```

implementation using Jubjub. However, none of these implementations is currently
compatible with Bitcoin, which uses secp256k1 and, if/when Taproot activates,

> BIP-340 Schnorr signatures. My plan is to implement a Bitcoin compatible FROST

Reviewer: ```suggestion
BIP-340 Schnorr signatures. My plan is to implement a Bitcoin-compatible FROST
```

> in C built on [ElementsProject/secp256k1-zkp][secp256k1-zkp github].

Reviewer: "and, if/when Taproot activates, BIP-340 Schnorr signatures" is a bit difficult to parse. Perhaps just remove it. BIP 340 Schnorr signatures also use secp256k1.

With FROST, a single Schnorr signature is generated that is indistinguishable
from a signature that is generated by a single party. An outside observer,
solely by analyzing the data in the blockchain, cannot detect that a multi-party
scheme was used to generate the signature. This makes FROST favorable from a

> scaling perspective as well because regardless of how many parties, and how

Reviewer: "as well" as what?

```suggestion
scaling perspective because regardless of how many parties, and how
```

> scaling perspective as well because regardless of how many parties, and how
> complex the hierarchy, the blockchain only stores a single signature that is

Reviewer: You haven't introduced what you mean by 'hierarchy'. Perhaps just drop it:

```suggestion
scaling perspective as well because regardless of how many parties were
involved in signature generation, the blockchain only stores a single signature that is
```

always the same size. In this way, privacy and scalability work hand in hand.

I’m currently in the early stages of the project and I’m focusing on gathering
feedback and drafting specifications. I will be sharing a detailed roadmap and
timeline soon. Follow [Brink][brink twitter] and [subscribe to the
newsletter][brink newsletter signup] for updates, and follow me on
[GitHub][jesse github] to track my progress!

[ln]: https://lightning.network

> [dlc]: https://github.com/discreetlogcontracts/dlcspecs/blob/master/Introduction.md

Reviewer: Alternative: https://bitcoinops.org/en/topics/discreet-log-contracts/

> [bip341]: https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki

Reviewer: Alternative: https://bitcoinops.org/en/topics/taproot/

(I'm biased but I think the optech topics pages are the best succinct summaries of Bitcoin topics :slightly_smiling_face:)
